sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java

```diff
@@ -723,84 +666,81 @@ public Duration getMaxReadTime() {
 
       private class PubsubReader extends DoFn<Void, T> {
         private static final int DEFAULT_PULL_SIZE = 100;
+        private static final int ACK_TIMEOUT_SEC = 60;
 
         @Override
         public void processElement(ProcessContext c) throws IOException {
-          Pubsub pubsubClient =
-              Transport.newPubsubClient(c.getPipelineOptions().as(PubsubOptions.class))
-                  .build();
-
-          String subscription;
-          if (getSubscription() == null) {
-            String topic = getTopic().asPath();
-            String[] split = topic.split("/");
-            subscription =
-                "projects/" + split[1] + "/subscriptions/" + split[3] + "_dataflow_"
-                + new Random().nextLong();
-            Subscription subInfo = new Subscription().setAckDeadlineSeconds(60).setTopic(topic);
-            try {
-              pubsubClient.projects().subscriptions().create(subscription, subInfo).execute();
-            } catch (Exception e) {
-              throw new RuntimeException("Failed to create subscription: ", e);
+          try (PubsubClient pubsubClient =
+                   PubsubClient.newClient(TransportType.APIARY,
+                       timestampLabel, idLabel, c.getPipelineOptions().as(PubsubOptions.class))) {
```

Comment:
it would be cool if you could also use try-with-resources to handle cleanup of the actual pubsub subscriptions (if topic, rather than subscription) is provided.